### ISSUE TITLE
UnboundLocalError: local variable 'copycmd' referenced before assignment

### ISSUE THREAD

**Issue by Infiniverse** [austinhyde/ansible-sshjail#16]
I'm trying the ansible-sshjail connection plugin against ansible 2.3 (with the dev_ansible_v2 branch).

I'm using 'jail-me' on the jail host.

When I try to run a simple command I get an error which appears to be code not environment.

```
joe:ansible-plays joe$ ansible -vvv -i inventory/devel/inventory all --limit=foo -m command -a pwd
No config file found; using defaults
META: ran handlers
Using module file /Users/joe/Documents/TrueSpeed/repositories/ansible/lib/ansible/modules/commands/command.py
<foo@10.17.0.2> ESTABLISH SSH CONNECTION FOR USER: None
<10.17.0.2> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/Users/joe/.ansible/cp/367f5a9c1f -tt 10.17.0.2 'which -s jailme'
<10.17.0.2> (0, '', 'Shared connection to 10.17.0.2 closed.\r\n')
<foo@10.17.0.2> ESTABLISH SSH CONNECTION FOR USER: None
<10.17.0.2> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/Users/joe/.ansible/cp/367f5a9c1f -tt 10.17.0.2 'jls -q jid name host.hostname path'
<10.17.0.2> (0, '1 ioc-94ea63de-1ee7-11e7-b2a8-080027ca2b91 foo /iocage/jails/94ea63de-1ee7-11e7-b2a8-080027ca2b91/root\r\n2 ioc-c5957792-1e72-11e7-ac07-080027ca2b91 asterisk-customer2 /iocage/jails/c5957792-1e72-11e7-ac07-080027ca2b91/root\r\n', 'Shared connection to 10.17.0.2 closed.\r\n')
<foo@10.17.0.2> ESTABLISH SSH CONNECTION FOR USER: None
<10.17.0.2> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/Users/joe/.ansible/cp/367f5a9c1f -tt 10.17.0.2 'jailme 1 /bin/sh -c '"'"'/bin/sh -c '"'"'"'"'"'"'"'"'echo ~'"'"'"'"'"'"'"'"''"'"' && sleep 0'
<10.17.0.2> (0, '/home/joe\r\n', 'Shared connection to 10.17.0.2 closed.\r\n')
<foo@10.17.0.2> ESTABLISH SSH CONNECTION FOR USER: None
<10.17.0.2> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/Users/joe/.ansible/cp/367f5a9c1f -tt 10.17.0.2 'jailme 1 /bin/sh -c '"'"'/bin/sh -c '"'"'"'"'"'"'"'"'( umask 77 && mkdir -p "` echo /home/joe/.ansible/tmp/ansible-tmp-1492681710.9-55117835483012 `" && echo ansible-tmp-1492681710.9-55117835483012="` echo /home/joe/.ansible/tmp/ansible-tmp-1492681710.9-55117835483012 `" )'"'"'"'"'"'"'"'"''"'"' && sleep 0'
<10.17.0.2> (0, 'ansible-tmp-1492681710.9-55117835483012=/home/joe/.ansible/tmp/ansible-tmp-1492681710.9-55117835483012\r\n', 'Shared connection to 10.17.0.2 closed.\r\n')
<foo@10.17.0.2> ESTABLISH SSH CONNECTION FOR USER: None
<10.17.0.2> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/Users/joe/.ansible/cp/367f5a9c1f -tt 10.17.0.2 mktemp
<10.17.0.2> (0, '/tmp/tmp.LXrU8SDe\r\n', 'Shared connection to 10.17.0.2 closed.\r\n')
<foo@10.17.0.2> ESTABLISH SSH CONNECTION FOR USER: None
<10.17.0.2> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/Users/joe/.ansible/cp/367f5a9c1f -tt 10.17.0.2 'chmod 0644 /tmp/tmp.LXrU8SDe'
<10.17.0.2> (0, '', 'Shared connection to 10.17.0.2 closed.\r\n')
<10.17.0.2> PUT /var/folders/p4/rjqvjlbx2rz1cs9x74f0tb6m0000gp/T/tmpQ5tjGA TO /tmp/tmp.LXrU8SDe
<10.17.0.2> SSH: EXEC sftp -b - -C -o ControlMaster=auto -o ControlPersist=60s -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o ConnectTimeout=10 -o ControlPath=/Users/joe/.ansible/cp/367f5a9c1f '[10.17.0.2]'
<10.17.0.2> (0, 'sftp> put /var/folders/p4/rjqvjlbx2rz1cs9x74f0tb6m0000gp/T/tmpQ5tjGA /tmp/tmp.LXrU8SDe\n', '')
<foo@10.17.0.2> REMOTE COPY /tmp/tmp.LXrU8SDe TO /iocage/jails/94ea63de-1ee7-11e7-b2a8-080027ca2b91/root/home/joe/.ansible/tmp/ansible-tmp-1492681710.9-55117835483012/command.py
The full traceback is:
Traceback (most recent call last):
  File "/Users/joe/Documents/TrueSpeed/repositories/ansible/lib/ansible/executor/task_executor.py", line 125, in run
    res = self._execute()
  File "/Users/joe/Documents/TrueSpeed/repositories/ansible/lib/ansible/executor/task_executor.py", line 521, in _execute
    result = self._handler.run(task_vars=variables)
  File "/Users/joe/Documents/TrueSpeed/repositories/ansible/lib/ansible/plugins/action/normal.py", line 45, in run
    results = merge_hash(results, self._execute_module(tmp=tmp, task_vars=task_vars, wrap_async=wrap_async))
  File "/Users/joe/Documents/TrueSpeed/repositories/ansible/lib/ansible/plugins/action/__init__.py", line 658, in _execute_module
    self._transfer_data(remote_module_path, module_data)
  File "/Users/joe/Documents/TrueSpeed/repositories/ansible/lib/ansible/plugins/action/__init__.py", line 342, in _transfer_data
    self._transfer_file(afile, remote_path)
  File "/Users/joe/Documents/TrueSpeed/repositories/ansible/lib/ansible/plugins/action/__init__.py", line 319, in _transfer_file
    self._connection.put_file(local_path, remote_path)
  File "/Users/joe/Documents/TrueSpeed/repositories/true-services-dev/ansible-plays/connection_plugins/sshjail.py", line 170, in put_file
    self._copy_file(tmp, out_path)
  File "/Users/joe/Documents/TrueSpeed/repositories/true-services-dev/ansible-plays/connection_plugins/sshjail.py", line 143, in _copy_file
    code, stdout, stderr = self._jailhost_command(copycmd)
UnboundLocalError: local variable 'copycmd' referenced before assignment

foo | FAILED | rc=-1 >>
Unexpected failure during module execution.
```

**Comment by fdutheil** [comment]
Hi Infiniverse.
I updated to ansible v2.3 on a Ubuntu control machine and used it without issue (my whole playbooks and the example you give):
```
$ ansible all --limit=test-jail -i inventory.pub -m command -a pwd
test-jail | SUCCESS | rc=0 >>
/

$ ansible --version
ansible 2.3.0.0
  config file = /home/florent/.ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.6 (default, Oct 26 2016, 20:30:19) [GCC 4.8.4]

```

I see you're using MacOS on your control machine, so it may be worth investigate on the python side (I used a macbook years ago, it was a recurrent issue). Try to add more information about your setup, like versions of OSes, of python on your host, of python in your jails, etc.

**Comment by Infiniverse** [comment]
Thanks for checking that!

On the host (mac):
```
ansible 2.3.0.0 (stable-2.3 b9ee0c5477) last updated 2017/04/20 09:13:46 (GMT +100)
  config file = 
  configured module search path = Default w/o overrides
  python version = 2.7.12 (default, Jun 29 2016, 14:05:02) [GCC 4.2.1 Compatible Apple LLVM 7.3.0 (clang-703.0.31)]
```

**Comment by Infiniverse** [comment]
Looks like copycmd isn't being set to anying... I'll try and work out why.

**Comment by Infiniverse** [comment]
Looks like the problem is here:

```
    def _copy_file(self, from_file, to_file):
        if self._play_context.become:
            # display.debug("_low_level_execute_command(): using become for this command")
            copycmd = self._play_context.make_become_cmd(' '.join(['cp', from_file, to_file]))

        display.vvv(u"REMOTE COPY {0} TO {1}".format(from_file, to_file), host=self.inventory_hostname)
        code, stdout, stderr = self._jailhost_command(copycmd)
        if code != 0:
            raise AnsibleError("failed to move file from %s to %s:\n%s\n%s" % (from_file, to_file, stdout, stderr))
```

If `self._play_context.become` is not set then `copycmd` isn't set and we get the errror.

I thought that with `jailme` we didn't need to become root.

**Comment by Infiniverse** [comment]
I've raised a pull request to fix this: #17

**Comment by austinhyde** [comment]
Fixed by #17
